### PR TITLE
RM-9250 Implement support for .NET DateOnly in ObjectBinding

### PR DESCRIPTION
**Pull request by a-ctor**
https://re-motion.atlassian.net/browse/RM-9250

### CODE REVIEW

File: Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocDateTimeValueTest.cs [hunk 1]
```diff
@@ -142,6 +142,17 @@ public void SetValueToNullableDateTimeNull ()
...
 
+#if NET6_0_OR_GREATER
+    [Test]
+    public void IBusinessObjectBoundControl_SetValueToDateOnly_ThrowsArgumentException ()
+    {
+      DateOnly dateOnly = new DateOnly(2006, 1, 1);
+      _bocDateTimeValue.IsDirty = false;
+      Assert.That(() => { ((IBusinessObjectBoundControl)_bocDateTimeValue).Value = dateOnly;},
+          Throws.InstanceOf<ArgumentException>()
+              .With.Message.EqualTo("Parameter 'value' has type 'System.DateOnly' when type 'System.Nullable`1[System.DateTime]' was expected. (Parameter 'value')"));
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Web.UnitTests/UI/Controls/BocDateTimeValueTest.cs:153]: Please with to the dedicated ArgumentException assertion:
`Throws.ArgumentException.With.ArgumentExceptionMessageEqualTo(messsage, parametername)`

File: Remotion/ObjectBinding/UnitTests/BusinessObjectStringFormatterServiceTests/GetPropertyString_WithDateTimeProperty.cs [hunk 2]
```diff
@@ -71,6 +71,25 @@ public void Scalar_WithValu_AndDateProperty ()
...
+      {
+        string actual = _stringFormatterService.GetPropertyString(_mockBusinessObject.Object, _mockProperty.Object, null);
+
+        _mockBusinessObject.Verify();
+        _mockProperty.Verify();
+        Assert.That(actual, Is.EqualTo("17.06.2000"));
+      }
+    }
+#endif
+
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/UnitTests/BusinessObjectStringFormatterServiceTests/GetPropertyString_WithDateTimeProperty.cs:92]: Turns out, we can revert this since the GetProperty should return DateTime, not DateOnly. Since we are mocking the property, there is no differernce between DateTimeProperty, DateProperty and DateOnlyProperty.

In the orignal specification, the assumption was we are using actual properties loaded via reflection from the domain in the test.

File: Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs [hunk 3]
```diff
@@ -51,5 +52,40 @@ public void Initialize_DateTimeProperty ()
...
+    [Test]
+    public void ConvertToNativePropertyType_DateOnlyProperty ()
+    {
+      var property = new DateOnlyProperty(
+          GetPropertyParameters(GetPropertyInfo(typeof(ClassWithAllDataTypes), propertyName: "DateOnly"), _businessObjectProvider));
+
+      var result = property.ConvertToNativePropertyType(new DateTime(2042, 4, 2, 13, 34, 20));
+
+      Assert.That(result, Is.InstanceOf<DateOnly>());
+      Assert.That(result!.As<DateOnly>(), Is.EqualTo(new DateOnly(2042, 4, 2)));
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs:87]:
```suggestion
      Assert.That(result, Is.EqualTo(new DateOnly(2042, 4, 2)));
```

File: Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs [hunk 4]
```diff
@@ -51,5 +52,40 @@ public void Initialize_DateTimeProperty ()
...
+    [Test]
+    public void ConvertFromNativePropertyType_DateOnlyProperty ()
+    {
+      var property = new DateOnlyProperty(
+          GetPropertyParameters(GetPropertyInfo(typeof(ClassWithAllDataTypes), propertyName: "DateOnly"), _businessObjectProvider));
+
+      var result = property.ConvertFromNativePropertyType(new DateOnly(2042, 4, 2));
+
+      Assert.That(result, Is.InstanceOf<DateTime>());
+      Assert.That(result!.As<DateTime>(), Is.EqualTo(new DateTime(2042, 4, 2)));
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs:75]:
```suggestion
      Assert.That(result, Is.EqualTo(new DateTime(2042, 4, 2)));
```

File: Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs [hunk 5]
```diff
@@ -16,6 +16,7 @@
 // 
 using System;
 using NUnit.Framework;
+using Remotion.Development.UnitTesting;
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs:19]:
```suggestion
```

File: Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs [hunk 6]
```diff
@@ -51,5 +52,40 @@ public void Initialize_DateTimeProperty ()
...
+#if NET6_0_OR_GREATER
+    [Test]
+    public void Initialize_DateOnlyProperty ()
+    {
+      IBusinessObjectDateTimeProperty property = new DateOnlyProperty(
+          GetPropertyParameters(GetPropertyInfo(typeof(ClassWithAllDataTypes), propertyName: "DateOnly"), _businessObjectProvider));
+
+      Assert.That(property.Type, Is.EqualTo(DateTimeType.Date));
+    }
+
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/UnitTests/BindableObject/DateTimePropertyTest.cs:65]: please add tests for NULL values (to and from native value)

File: Remotion/ObjectBinding/Core/BindableObject/Properties/DateOnlyProperty.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    public DateOnlyProperty (Parameters parameters)
+        : base(parameters)
+    {
+    }
+
+    public override object? ConvertFromNativePropertyType (object? nativeValue)
+    {
+      if (nativeValue != null)
+        return ((DateOnly)nativeValue).ToDateTime(TimeOnly.MinValue);
+      return base.ConvertFromNativePropertyType(nativeValue);
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Core/BindableObject/Properties/DateOnlyProperty.cs:19]: the other proprties use
```suggestion
      var dateOnly = ArgumentUtility.CheckType<DateOnly?>("nativeValue", nativeValue);
      if (dateOnly != null)
        return (dateOnly.ToDateTime(TimeOnly.MinValue);
      return base.ConvertFromNativePropertyType(dateOnly);
```

File: Remotion/ObjectBinding/Core/BindableObject/Properties/DateOnlyProperty.cs [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    }
+
+    public override object? ConvertToNativePropertyType (object? publicValue)
+    {
+      if (publicValue != null)
+        return DateOnly.FromDateTime((DateTime)publicValue);
+      return base.ConvertToNativePropertyType(publicValue);
```
Comment (MichaelKetting) [inline Remotion/ObjectBinding/Core/BindableObject/Properties/DateOnlyProperty.cs:26]: The other properties use
```suggestion
      var dateTime = ArgumentUtility.CheckType<DateTime?>("publicValue", publicValue);
      if (dateTime != null)
        return DateOnly.FromDateTime(dateTime );
      return base.ConvertToNativePropertyType(dateTime);
```

**Review by MichaelKetting — CHANGES_REQUESTED** [review]

**Comment by a-ctor** [comment]
@MichaelKetting I have addressed all the review comments but I haven't had the time to fix the screenshot tests. These need to be fixed for the build to succeed. Apart from that the PR is done
